sqlalchemy-stubs/sql/sqltypes.pyi

```diff
@@ -3,78 +3,104 @@ from datetime import datetime
-    comparator_factory: Any = ...
-
-class Concatenable:
-    class Comparator(TypeEngine.Comparator): ...
-    comparator_factory: Any = ...
-
-class Indexable:
-    class Comparator(TypeEngine.Comparator):
-        def __getitem__(self, index: Any): ...
-    comparator_factory: Any = ...
-
-class String(Concatenable, TypeEngine[str]):
-    __visit_name__: str = ...
-    RETURNS_UNICODE: Any = ...
-    RETURNS_BYTES: Any = ...
-    RETURNS_CONDITIONAL: Any = ...
-    RETURNS_UNKNOWN: Any = ...
-    length: Any = ...
-    collation: Any = ...
+_TE = TypeVar("_TE", bound=TypeEngine[Any])
+_ST = TypeVar("_ST", bound=SchemaType)
+_EN = TypeVar("_EN", bound=Enum)
+
+_NumericType = Union[float, Decimal]
+_JSONType = Union[str, Mapping[Any, Any], List[Any]]
+
+class _LookupExpressionAdapter(Generic[_T]):
+    class Comparator(TypeEngine.Comparator[_TE]): ...
+    comparator_factory: Type[Comparator[TypeEngine[_T]]] = ...
+
+class Concatenable(Generic[_T]):
+    class Comparator(TypeEngine.Comparator[_TE]): ...
+    comparator_factory: Type[Comparator[TypeEngine[_T]]] = ...
+
+class Indexable(Generic[_T]):
+    class Comparator(TypeEngine.Comparator[_TE]):
+        def __getitem__(self, index: Any) -> Any: ...
+    comparator_factory: Type[Comparator[TypeEngine[_T]]] = ...
+
+class String(Concatenable[str], TypeEngine[str]):  # type: ignore[misc]
+    __visit_name__: str = ...
+    RETURNS_UNICODE: ClassVar[langhelpers._symbol] = ...
+    RETURNS_BYTES: ClassVar[langhelpers._symbol] = ...
+    RETURNS_CONDITIONAL: ClassVar[langhelpers._symbol] = ...
+    RETURNS_UNKNOWN: ClassVar[langhelpers._symbol] = ...
+    length: Optional[int] = ...
+    collation: Optional[str] = ...
+    @overload
     def __init__(
         self,
-        length: Optional[Any] = ...,
-        collation: Optional[Any] = ...,
+        length: Optional[int] = ...,
+        collation: Optional[str] = ...,
         convert_unicode: bool = ...,
-        unicode_error: Optional[Any] = ...,
         _warn_on_bytestring: bool = ...,
         _expect_unicode: bool = ...,
     ) -> None: ...
-    def literal_processor(self, dialect: Any): ...
-    def bind_processor(self, dialect: Any): ...
-    def result_processor(self, dialect: Any, coltype: Any): ...
+    @overload
+    def __init__(
+        self,
+        length: Optional[int],
+        collation: Optional[str],
+        convert_unicode: Literal["force"],
+        unicode_error: str,
+        _warn_on_bytestring: bool = ...,
+        _expect_unicode: bool = ...,
+    ) -> None: ...
+    @overload
+    def __init__(
+        self,
+        length: Optional[int],
+        *,
+        convert_unicode: Literal["force"],
+        unicode_error: str,
+        _warn_on_bytestring: bool = ...,
+        _expect_unicode: bool = ...,
+    ) -> None: ...
+    def literal_processor(self, dialect: Any) -> _LiteralProcessor[str]: ...
+    def bind_processor(
+        self, dialect: Any
+    ) -> Optional[_BindProcessor[str]]: ...
```

Comment:
so I noticed in type_api that `_BindProcessor` is a class with a ``__call__`` method, and isn't just plain `typing.Callable`.  How does this work when a plain Python callable is passed somewhere to bind_processor(), and it's not actually a subclass of `_BindProcessor`, which doesn't even exist in the library?      shouldn't mypy throw an error on that ?

Reply:
`_BindProcessor`, `_LiteralProcessor`, and `_ResultProcessor` are all [callback protocols](https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols). This means any callable that matches the signature (minus `self`) of `__call__` will match that protocol. I tend to use callback protocols over `typing.Callable` because they are generally more expressive (for instance, you can define keyword arguments in callback protocols).